### PR TITLE
timeval: use mach time on MacOS if necessary

### PR DESCRIPTION
**Pull request by dtikhonov**
If clock_gettime() is not supported, one can use mach_absolute_time().

### CODE REVIEW

File: lib/timeval.c [hunk 1]
```diff
@@ -84,6 +84,37 @@ struct curltime curlx_tvnow(void)
   return cnow;
 }
 
+#elif defined(__APPLE__)
```
Comment (nickzman) [inline lib/timeval.c]: Wouldn't it be better to just check for mach_absolute_time() in the configure script?

Comment (dtikhonov) [reply lib/timeval.c]: Probably.

Comment (dtikhonov) [reply lib/timeval.c]: There are two ways to configure _curl_: configure and cmake.  Are pull requests expected to update both?

Comment (bagder) [reply lib/timeval.c]: Preferred at least! Just adding a check for a function is pretty easy in both though.

File: lib/timeval.c [hunk 2]
```diff
@@ -84,6 +84,37 @@ struct curltime curlx_tvnow(void)
...
+
+#include <stdint.h>
+#include <mach/mach_time.h>
+
+struct curltime curlx_tvnow(void)
```
Comment (bagder) [inline lib/timeval.c]: Also beware that I sneakily just renamed that function to `Curl_now()` since 5d543fe906 =)

File: lib/timeval.c (hunk 1 above)
Comment (dtikhonov) [reply lib/timeval.c]: Done!  Well, except that it does not work with cmake.

**Comment by bagder** [comment]
Thanks!

First, the CI error, as I mentioned above the function has changed name:
~~~
timeval.c:92:17: error: no previous prototype for function 'curlx_tvnow' [-Werror,-Wmissing-prototypes]
struct curltime curlx_tvnow(void)
~~~

Do you really need that complicated configure logic to detect the function? Doesn't this work?
~~~diff
diff --git a/configure.ac b/configure.ac
index 68b3a071a..5272feaa0 100755
--- a/configure.ac
+++ b/configure.ac
@@ -3383,10 +3383,11 @@ AC_CHECK_FUNCS([geteuid \
   getpwuid \
   getpwuid_r \
   getrlimit \
   gettimeofday \
   if_nametoindex \
+  mach_absolute_time \
   pipe \
   setlocale \
   setmode \
   setrlimit \
   utime \
~~~

And the corresponding fix for cmake:
~~~diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 3bfb5febd..6b555b918 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -876,10 +876,11 @@ check_symbol_exists(getrlimit      "${CURL_INCLUDES}" HAVE_GETRLIMIT)
 check_symbol_exists(setlocale      "${CURL_INCLUDES}" HAVE_SETLOCALE)
 check_symbol_exists(setrlimit      "${CURL_INCLUDES}" HAVE_SETRLIMIT)
 check_symbol_exists(fcntl          "${CURL_INCLUDES}" HAVE_FCNTL)
 check_symbol_exists(ioctl          "${CURL_INCLUDES}" HAVE_IOCTL)
 check_symbol_exists(setsockopt     "${CURL_INCLUDES}" HAVE_SETSOCKOPT)
+check_symbol_exists(mach_absolute_time "${CURL_INCLUDES}" HAVE_MACH_ABSOLUTE_TIME)
 
 # symbol exists in win32, but function does not.
 if(WIN32)
   if(ENABLE_INET_PTON)
     check_function_exists(inet_pton HAVE_INET_PTON)
~~~

**Comment by dtikhonov** [comment]
Thanks for the pointers, @badger!  I have little experience with configure scripts -- what I have in the pull request is a careful copy-paste-modify job, and it still took me much longer than the actual code.  I will give this another shot in the next couple of days.

**Comment by dtikhonov** [comment]
Will do another branch

**Pull request closed without merging**